### ISSUE TITLE
proxy settings Client>https>reverseproxy>https>snipeit - Apache - not working

### ISSUE THREAD

**Issue by avielc** [grokability/snipe-it#10644]
### Debug mode

- [X] I have enabled debug mode
- [X] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

### Describe the bug

Trying to set up reverse proxy apache here
I'm being stomped by multiple errors
First on list: 

`Access to font at 'https://localhost/css/fonts/fontawesome-webfont.ttf?v=4.7.0' from origin 'https://proxyserver' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`
there are about 4 fonts failing. and all permissions are set on 755. 

Second (this might be on me) 
Whenever I click any link I am just redirected to the localhost address instead of having the Proxy remain as is and just pull information from the local host. 

Adding some parts of code here, hoping anyone with experience can catch my error with a glimpse

`        SSLEngine on
        SSLProxyEngine on
        SSLProxyCheckPeerCN off
        SSLProxyCheckPeerName off
        SSLProxyCheckPeerExpire off
        SSLCertificateFile <It's here>
        SSLCACertificateFile <It's here>
        ProxyTimeout 2700
        SSLProtocol All -SSLv2 -SSLv3
        SSLHonorCipherOrder On
        SetEnvIf Origin "https://inventory.internal.domain/$" AccessControlAllowOrigin=$0
        Header always set Access-Control-Allow-Origin %{AccessControlAllowOrigin}e env=AccessControlAllowOrigin
        Header always set Access-Control-Allow-Credentials true
        Header always set Access-Control-Allow-Headers "Origin, Authorization, Accept"
        Header always set Access-Control-Allow-Methods "POST, GET, OPTIONS, DELETE, PUT"
        Header always set Access-Control-Max-Age "600"
        Header always set Access-Control-Expose-Headers "Content-Security-Policy, Location"
        Header merge Vary Origin
    RewriteEngine On
    RewriteCond %{REQUEST_METHOD} OPTIONS
    RewriteRule ^(.*)$ $1 [R=204,L]
        ProxyPreserveHost off
        ProxyPass / https://localhost/ nocanon
        ProxyPassReverse / https://localhost/
        ProxyRequests Off
        RequestHeader set X-Forwarded-Proto "https"
        RequestHeader set X-Forwarded-Port "<randomport on proxy>"

`

Thanks

### Reproduction steps

1.basically just create an apache reverse proxy with ssl certificate
2.
3.
...

### Expected behavior

just to see the snipe system through a proxy, not exposing the server to the world

### Snipe-IT Version

5.3.8

### Operating System

centos

### Web Server

apache on proxy, nginx on snipe

### PHP Version

7.4 on nginx

### Error messages

```shell
Additional error on sniper server as well as on Apache

Error with Feature-Policy header: Unrecognized feature: 'ambient-light-sensor'.
Error with Feature-Policy header: Unrecognized feature: 'animations'.
Error with Feature-Policy header: Unrecognized feature: 'battery'.
Error with Feature-Policy header: Unrecognized feature: 'legacy-image-formats'.
Error with Feature-Policy header: Unrecognized feature: 'oversized-images'.
Error with Feature-Policy header: Unrecognized feature: 'publickey-credentials'.
Error with Feature-Policy header: Unrecognized feature: 'unsized-media'.
Error with Feature-Policy header: Unrecognized feature: 'vibrate'.
Error with Feature-Policy header: Unrecognized feature: 'wake-lock'.
```

**Comment by avielc** [comment]
Thanks for the assignment. 
I'd really wish there was some clearer info about Apache\Nginx config example so I can replicate it and adjust it to my needs. 
So I won't repeat my issues, to the length, just sum it up (and ask me to elaborate if you need more info) 
1. some UI doesn't appear and show the 404 error on some Fonts files. 
2. clicking any link will force the website to change into the internal address instead of allowing the reverseproxy to be the middle man.

**Comment by uberbrady** [comment]
What do you have your `APP_URL` in your `.env` set to - does it match what you see in the URL Bar at the top of your screen?

You can also try (but I don't recommend, maybe just use for testing) setting `APP_ALLOW_INSECURE_HOSTS` to `true` in your `.env` - but really, you should try and fix your `APP_URL` if you can.

**Comment by avielc** [comment]
Thanks for replying man!
the APP_URL is set to the internal address, I thought the Apache reverse proxy should compensate on links that match the internal address and just auto complete it on the external address, no? 
e,.g : https://snipe.internal.domain.com <-----Reverse proxy to---> https://snipeit.externaladdress.com

**Comment by snipe** [comment]
The `APP_URL` should always be what you see in your URL bar, [per the documentation](https://snipe-it.readme.io/docs/configuration#required-basic-app-settings). 

<img width="815" alt="Screen Shot 2022-02-11 at 4 55 40 PM" src="https://user-images.githubusercontent.com/197404/153689481-cf976efb-d1cb-4a36-b031-33eca360f859.png">

**Comment by uberbrady** [comment]
The `APP_URL` gets used in a couple of different places - one is that when the system is trying to figure out what the hostname is in order to give fully-qualified URL's for various things. Another is for URL's in automated emails.

As to what your proxy is doing - that I couldn't quite tell you, but I'm still pretty sure that setting your `APP_URL` to `https://snipeit.externaladdress.com` ought to be the right way to go? Let us know if that helps.

**Comment by avielc** [comment]
Thanks for the info to both of you. 
Is it possible to set it to Null? 
I would very much like to have the APP_URL be flexible if I need to use it for internal while at work, and for exceptional cases through the reverse proxy.

**Comment by snipe** [comment]
Per the documentation (and the screen shot I just pasted you above), it should not be set to null.

**Comment by avielc** [comment]
@snipe - and currently , it is set to the internal domain address.
I understand the documentation quote you mentioned, I am looking for a solution here. 
Can you please help me understand how we can achieve a solution where it can be reached on both addresses properly?

**Comment by snipe** [comment]
It unfortunately isn't possible to use two URLs. As @uberbrady mentioned, that `APP_URL` is used to hardcode links throughout the system and in various emails. I cannot have two values at once.

**Comment by avielc** [comment]
and there's absolutely no way to fool it using reverse proxy match? :-\ (asking to confirm, not trying to be annoying here)

**Comment by snipe** [comment]
I suppose that would depend on how you configure your reverse proxy, but the `APP_URL` link itself would have to be accessible via whatever the single `APP_URL` is that you use.

**Comment by snipe** [comment]
Did you try the `APP_ALLOW_INSECURE_HOSTS` config setting @uberbrady [suggested before](https://github.com/snipe/snipe-it/issues/10644#issuecomment-1036867164)? As he mentioned, that should NOT be the solution here, as it's likely to break in versions moving forward, but it might work for you for now.

**Comment by avielc** [comment]
that's obviously true. (having the reverse proxy being able to access that internal machine) 
That's why I would like to know if there's a way to configure the reverse proxy to bypass that hard limit. 
I've tried all kinds of play, but like you said, it seems the javascript and some other UI \ the link it self breaks. 

APP_ALLOW_INSECURE_HOSTS - didn't see that config.  (I'll look it up on Sunday) 
I did however try to play with the APP_TRUSTED_PROXIES - don't think it helped here at all. (nothing happened with this)

**Comment by snipe** [comment]
Unfortunately, configuring your reverse proxy and working our hair pinning is outside the scope of Snipe-IT support. Let us know how you make out with that config setting.

**Comment by avielc** [comment]
(checked the documentations now for what that means. 
You make it sound like a really risky option. 
I might as well set the APP_URL to the proxy address 

I did kinda hope you could have some generic config available, don't really need something tailored outside your reverseproxy page - as you wrote in the reverse proxy page. What did you mean by that?
![image](https://user-images.githubusercontent.com/13572969/153690983-cf55c879-ea64-415b-9a97-a80b6f4e0b61.png)

**Comment by snipe** [comment]
We do not have generic configs for this stuff, since everyone else's setup is slightly different.

**Comment by avielc** [comment]
Fair point, However,  you did mention the Apache\Nginx header to pass on and the APP_TRUSTED_PROXIES.
did you test these scenario's mentioned above? 
is there any documentation on what was tested? 
Perhaps I can figure a proper configuration out of what you used. If you allow it of course.

**Comment by snipe** [comment]
We don't use proxies with Snipe-IT, so we don't have anything to offer you. It was tested with individual user setups back when it was written originally, which was years ago.

**Comment by avielc** [comment]
Gotcha, 
Well thanks again for all the help, and sorry to have bothered with such rare case .
Closing this issue now

**Comment by snipe** [comment]
No problem - do let us know if that additional env var setting helps get you closer.

**Comment by avielc** [comment]
Will do!

**Comment by avielc** [comment]
As promised - here's an update 
@snipe @uberbrady - adding that flag (didn't exist in the `.env` at all.) made the page appear properly without any issues. 
I've secured it behind reverse proxy with password\ip filtering. - both internal and reverse proxy page use SSL Certificates and HTTPS access only. 
However You did worry me with it being not secured and shouldn't be exposed to the outside world, 
Any chance I can get some more info about it? I really want to understand the risks here and if it's best to avoid using it at all.

Thanks!

**Comment by snipe** [comment]
We don’t ever update the .env on your side. Any new .env vars have to be manually added by you based on the documentation, since .env is not checked into git. It’s a file you create.

**Comment by snipe** [comment]
(Brady is writing up some additional info on the hostname stuff.)

**Comment by uberbrady** [comment]
But in terms of ways to ensure that your server is actually secure, the main thing you need to do is restrict the hostnames that will answer for your site. This is usually done with name-based virtualhosts.

But what you *don't* want is, if a request comes across that has the wrong hostname, that you answer it.

So it sounds like you want two possible hostnames - the 'inside' hostname and the 'outside' hostname.

You can restrict that at the proxy layer, or at the webserver layer. So long as you do that, you should be OK.

We are probably going to have to adjust our messaging about that,

**Comment by avielc** [comment]
> We don’t ever update the .env on your side. Any new .env vars have to be manually added by you based on the documentation, since .env is not checked into git. It’s a file you create.

I thought it was available before, thank you for that! and I did notice .env doesn't get pulled from the git, which is of course great as it's a configuration file, obviously.



> But in terms of ways to ensure that your server is actually secure, the main thing you need to do is restrict the hostnames that will answer for your site. This is usually done with name-based virtualhosts.
> 
> But what you _don't_ want is, if a request comes across that has the wrong hostname, that you answer it.
> 
> So it sounds like you want two possible hostnames - the 'inside' hostname and the 'outside' hostname.
> 
> You can restrict that at the proxy layer, or at the webserver layer. So long as you do that, you should be OK.
> 
> We are probably going to have to adjust our messaging about that,

So just to try and translate what you said to "my language" (my understanding in all this isn't that great, sorry for that)
Basically Apache has "Servername" which is  has SSL Certificate to that name. 
it's proxymatch is redirecting to the internal name, which I added the IP translation under hosts, 
+ all the security of credentials (on the page before the snipe-it login page) \ ip filtering

All that should be secure enouogh? or did I misunderstand something from waht you said?

**Comment by avielc** [comment]
Hi Guys
Sorry for the reopn, not sure what has changed (maybe due to latest version) 
the proxy and snipe fell back to the same issue where once I click on Login (reaching the Snipe login page) 
the address changes to the internal url (one that is specified in APP_URL) 
The flags are in place as they were, I didn't touch any configs. 
so both `APP_ALLOW_INSECURE_HOSTS=true` and `APP_TRUSTED_PROXIES=` are both there. 

so what am I missing? did something change in recent version (I updated just now to 6.10 in hope it'll resolve it) that ignores the flags?

**Comment by avielc** [comment]
> Did you try the `APP_ALLOW_INSECURE_HOSTS` config setting @uberbrady [suggested before](https://github.com/snipe/snipe-it/issues/10644#issuecomment-1036867164)? As he mentioned, that should NOT be the solution here, as it's likely to break in versions moving forward, but it might work for you for now.

Just adding to it. 
Any chance this was removed? is there anything to replace it?

**Comment by avielc** [comment]
one more thing to add:
Tried to login through a clean Incognito page
the first login forwarded to a second login on the server, 
If I tried to refresh the external proxy page  and login again, I receive the following message inside Snipe:
![image](https://user-images.githubusercontent.com/13572969/189160423-b7d58f72-ccfb-4846-b7ed-6c0e8eba385a.png)

Any ideas?

**Comment by avielc** [comment]
Putting it down in notes here so if someone runs into this issue it might help them. 

I'm not really sure what happened or how it got fixed, I ran caching a couple of times - which didn't help. 
last I did was change the `APP_URL` to the external  proxy, did `composer dump-autoload` (and cache clear,etc) 
and then saw that it got stuck on the proxy as intended. 
Changed back to the internal address and now Proxy is working again.
